### PR TITLE
Ignore new blocks when -stopatheight target has been reached

### LINKED ISSUE #13477: -stopatheight is imprecise
As of latest master (2b770080a49f282e51b7e4d176b2d9fbe0f8f5db).

```sh
src/bitcoind -testnet -stopatheight=100000

...

2018-06-15T12:49:59Z UpdateTip: new best=000000000086a985696447c96c8a3520dbe5536287ef9fbf5b220b66fdb684c7 height=100208 version=0x00000002 log2_work=54.043471 tx=599130 date='2013-08-11T14:57:43Z' progress=0.032961 cache=43.5MiB(302489txo)
2018-06-15T12:49:59Z UpdateTip: new best=000000000060c93d2cf7a5cd5ce510c31cf96b983ee64ec5c627cf2055249500 height=100209 version=0x00000002 log2_work=54.043562 tx=599131 date='2013-08-11T15:01:20Z' progress=0.032961 cache=43.5MiB(302490txo)
2018-06-15T12:49:59Z Shutdown: In progress...
2018-06-15T12:49:59Z addcon thread exit
2018-06-15T12:49:59Z net thread exit
2018-06-15T12:49:59Z UpdateTip: new best=000000000075b13f384b05ac740b776593dc4f53d2d92cf9c2a81475335a3cef height=100210 version=0x00000002 log2_work=54.043653 tx=599136 date='2013-08-11T15:09:42Z' progress=0.032961 cache=43.5MiB(302495txo)
2018-06-15T12:49:59Z msghand thread exit
2018-06-15T12:50:01Z opencon thread exit
2018-06-15T12:50:01Z scheduler thread interrupt
2018-06-15T12:50:01Z Dumped mempool: 7e-06s to copy, 0.00237s to dump
2018-06-15T12:50:07Z Shutdown: done
```

Having precise control can be useful for generating a deterministic blockchain snapshot. That in turn could be useful when bootstrapping a low performance device (though hopefully #9483 will make that unnecessary), as well as for validating the chain in parallel on multiple machines.

One easy solution might be to just roll back to the target.

### PR DESCRIPTION
**Pull request by jonasschnelli**
If the `-stopatheight` target has been reached, a shutdown will be triggered, though, already loaded blocks will still be activated, making it impossible to precisely stop at a certain height.

This PR will ignore future blocks once the `-stopatheight` target has been reached.

Fixes #13477

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -2705,6 +2705,12 @@ bool CChainState::ActivateBestChain(CValidationState &state, const CChainParams&
         {
             LOCK(cs_main);
             CBlockIndex* starting_tip = chainActive.Tip();
+
+            if (nStopAtHeight && starting_tip && starting_tip->nHeight >= nStopAtHeight && ShutdownRequested()) {
```
[the remark this answers is not in the document]
Comment (jonasschnelli) [reply src/validation.cpp]: The upper if statement with the `ShutdownRequested()` does make sure that on future calls of ActivateBestChain() and after `StartShutdown()` will not process blocks when stopheight is set at a shutdown is triggered.

The lower if statement is there to trigger the shutdown.

**Comment by fanquake** [comment]
Note previous discussion about `-stopatheight` in #13490.

**Comment by laanwj** [comment]
> Note previous discussion about -stopatheight in #13490.

I tend to agree with @sdaftuar's reasoning that it is undesirable to burden the validation code further with this.

Then again, if we're not going to fix this we should close #13477 (with "wontfix" and this reasoning), otherwise people will wasting time trying to fix it.

**Comment by Sjors** [comment]
This change is already a lot smaller than the previous attempt.

Being able to produce a deterministic UTXO set for a given height, as well as block files and indexes, could be quite useful. There may be other approaches than `-stopatheight`, or perhaps we can fix `-stopatheight` in a less disruptive way, e.g. by rolling back after overshooting.

**Comment by NicolasDorier** [comment]
@jonasschnelli this is not good.
Here is what I wished I could do:

```bash
bitcoind -stopatheight 500000
# Wait it reached it
bitcoin-cli gettxoutsetinfo
# Note the UTXOSet hash
bitcoin-cli stop
tar xcf "bitcoin-chainstate-snapshot-500000.tar" ~/.bitcoin/chainstate
# Upload the  bitcoin-chainstate-snapshot-500000.tar
# Sign the snapshot
```

Then somebody who wants to verify and sign the snapshot as well would repeat

```bash
bitcoind -stopatheight 500000
# Wait it reached it
bitcoin-cli gettxoutsetinfo
# Note the UTXOSet hash
```
Then try to import the UTXO set from the tar

```bash
tar xvf  bitcoin-chainstate-snapshot-500000.tar -C ~/.bitcoin/chainstate
bitcoind -stopatheight 500000
bitcoin-cli gettxoutsetinfo
# Check that the utxoset hash is the same as the one of your own synched machine
```
Then he can add his signature `bitcoin-chainstate-snapshot-500000.tar`.

If you stop the node once the height is reached, then there is no way to have the hash of the utxoset, and no way for somebody else to independently verify a given utxoset inside a tar file.

**Comment by Sjors** [comment]
@NicolasDorier I totally agree your workflow is what we should strive for. However this PR only fixes a bug in the existing behaviour. Let's redefine `-stopatheight` (e.g. by adding a `stop_sync_but_dont_exit` param) in a followup PR.

**Comment by NicolasDorier** [comment]
Ah got it, this is already the current situation.

File: src/validation.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (NicolasDorier) [reply src/validation.cpp]: What I don't understand is that if `ShutdownRequested()` is false, then you will end up processing one more block, and this is not what you want. Removing `ShutdownRequested()` and just call `StartShutdown` at this place?

**Pull request closed without merging**